File: remove-invalid.java
// Time, Space -  O(2^N), O(N)

class Solution {
    public List<String> removeInvalidParentheses(String s) {
        
        if(s == null) {
            return new ArrayList<>();
        }
        
        Queue<String> q = new LinkedList<>();
        
        List<String> result = new ArrayList<>();
        HashSet<String> set = new HashSet<>();
        q.add(s);
        set.add(s);
        
        Boolean flag = false;
        
        while(!q.isEmpty() && !flag) {
            int size = q.size();
            
            for(int i=0;i<size;i++) {
                String ele = q.poll();
            
                if(isValid(ele)) {
                    result.add(ele);
                    flag = true;
                }
                if(!flag) {
                    for(int j=0;j<ele.length();j++) {
                        if(Character.isLetter(ele.charAt(j))) {
                            continue;
                        }
                        String child = ele.substring(0,j) + ele.substring(j+1,ele.length());
                        if(!set.contains(child)) {
                            q.add(child);
                            set.add(child);
                        }
                        
                    }
                }
            }
        }
        
        return result;
    
    }
    
    private boolean isValid(String s) {
        if(s == null || s.length() == 0) {
            return true;
        }
        
        int cnt = 0;
        for(int i=0;i<s.length();i++) {
            char ch = s.charAt(i);
            if(Character.isLetter(ch)) {
                continue;
            }
            if(ch == '(') {
                cnt++;
            }
            else {
                cnt--;
            }
            if(cnt == -1) {
                return false;
            }
        }
        
        
        return cnt==0;
        
    }
}

File: removeInvalid.py
#Time: O(n^2)
#Space: O(n^2)
#Program ran on leetcode successfully

class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        res = []
        if s is None or len(s) == 0:
            return res
        q = []
        stringSet = set()
        q.append(s)
        stringSet.add(s)
        flag = False
        while (q and not flag):
            size = len(q)
            for i in range(size):
                tmp = q.pop(0)
                if (self.isValid(tmp)):
                    res.append(tmp)
                    flag = True
                if (not flag):
                    for j in range(len(tmp)):
                        c = tmp[i]
                        if (c.isalpha()):
                            continue
                        child = c[0:j-1] + c[j:]
                        if child not in stringSet:
                            q.append(child)
                            stringSet.add(child)
        
        return res
    
    def isValid(self, s: str):
        count = 0
        for i in range(len(s)):
            c = s[i]
            if c.isalpha(): 
                continue
            if c == '(': 
                count += 1
            elif c == ')':
                count += -1
            if count < 0:
                return False
        
        return count == 0
        
        